File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    phone_dict = {}
    
    phone_book.sort(key=len)
    min_length = len(phone_book[0])
    
    for phone_num in phone_book:
        min_length = min(min_length, len(phone_num))
        
    for phone_num in phone_book:
        idx = min_length - 1
        curr_partial = phone_num[:min_length]
        
        while idx < len(phone_num):
            if curr_partial in phone_dict:
                return False
            else:
                idx += 1
                if idx == len(phone_num): # 부분 집합을 끝까지 돌았음에도 접두어를 발견하지 못한 경우 -> 해당 번호도 dict에 추가
                    phone_dict[phone_num] = True
                    break
                curr_partial += phone_num[idx]
                    
    return True
        

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.js
function solution(phone_book) {
    const map = new Map()
    phone_book.sort((a, b) => a.length - b.length)
    
    const start = phone_book[0].length - 1
    for (const phone of phone_book) {
        let curr = phone.substring(0, start)
        for (let i = start; i < phone.length - 1; i++) {
            curr = curr + phone[i]
            
            if (map.get(curr)) {
                return false
            }
        }
        map.set(phone, true)
    }
    
    return true
}